enhancements/custom-alert-configuration.md

```diff
@@ -0,0 +1,438 @@
+by testing.
+
+### Drawbacks
+
+ValidatingAdmissionPolicy is still in alpha.
+
+
+## Design Details
+
+### Open Questions [optional]
+
+1. Are there any security implications with storing service accounts names in a
+  CRD and comparing these agains `request.userInfo.username` in the
+   `ValidatingAdmissionPolicy` ?
+
+### Test Plan
+
+- e2e test for kube-controller-manager operator should test a presence of
+  the alert and `ValidatingAdmissionPolicy` and that it guards modification
+  of PDBs labels. And correctly excludes labeled PDBs from
+  `PodDisruptionBudgetAtLimit` alert.
+
+### Graduation Criteria
+
+This should not be a user facing API.
+
+#### Dev Preview -> Tech Preview
+
+#### Tech Preview -> GA
+
+#### Removing a deprecated feature
+
+### Upgrade / Downgrade Strategy
+
+
+### Version Skew Strategy
+
+### Operational Aspects of API Extensions
+
+
+#### Failure Modes
+
+
+#### Support Procedures
+
+
+## Implementation History
+
+- 2023-01-30: Initial enhancement proposal
+
+
+## Alternatives
+
+### Use a validating admission webhook instead
+
+We could guard a PDB with validating admission webhook to prevent an unhandled
+labels change misuse. Based on the user info from AdmissionRequest we would
+decide if they can modify the label. We could check for either hardcoded
+user/SA that does the request or for KubeVirt alternatively check:
+
+- if the PDB is owned by the VM
+- do SubjectAccessReview if the user has access to the VM
+- check RBAC rights for some KubeVirt resource that would identify a controller
+
+These could be potential issue with this alternative:
+- less reliable than Validating Admission Policy
+- would increase network latency for a core object
+- maintenance burden ( would be probably owned by
+  kube-controller-manager-operator)
+
+### Introduce a new metric
+
+Mainly, KubeVirt should be the only one allowed to disable this alert. We could
+achieve this by introducing a custom metric called for example
+`ocp_kubevirt_vmi_owned_pdbs`. KubeVirt would track its own special PDBs that
+do not conform to the standard use case. We would then subtract this from the
+rest of the PDBs at limit to stop alerting on the KubeVirt ones.
+
+The metric should have at least a namespace and poddisruptionbudget name.
+
+The alert query could look like this:
+
+```json
+max by (namespace, poddisruptionbudget) (kube_poddisruptionbudget_status_current_healthy >= kube_poddisruptionbudget_status_desired_healthy and on (namespace, poddisruptionbudget) (kube_poddisruptionbudget_status_expected_pods > 0) unless on(namespace, poddisruptionbudget) (ocp_kubevirt_vmi_owned_pdbs))
+```
+
+Kubevirt would maintain/own `ocp_kubevirt_vmi_owned_pdbs` metric and ensure
+only the valid PDBs are ignored. This would make it more difficult for users
+to insert their own PDBs into this ignore list.
+
+The disadvantage is that it is not scalable for more components. The metric
+describes only KubeVirt use case.
+
+### Silencing the alert
```

Comment:
Another similar alternative would be to use [inhibition rules](https://prometheus.io/docs/alerting/latest/configuration/#inhibit_rule): KubeVirt could fire an info alert for each VM and there would be a generic inhibition rule that says "mute any PDB alert that have a matching KubeVirt info alert".
The main blocker is that inhibition rules are defined in the Alertmanager configuration so it would have to be . And the alert would still show up as firing in the OCP console.

Reply:
Thanks, I have added this as an option to the doc

Reply:
And yet another option to consider is the new [Custom Alerts feature](https://docs.openshift.com/container-platform/4.12/monitoring/managing-alerts.html#managing-core-platform-alerting-rules_managing-alerts). This is currently TP but we're looking to graduate this to GA for 4.13.
This has similar properties as silences and inhibitions (original alert is still present and fires).

Reply:
actually, this seems to work pretty well for kubevirt, after testing it. Will include it in alternatives shortly

Reply:
made a mistake in my setup - this seems it would just stop the notifications without affecting the rest

Reply:
added the option there as well

Reply:
I think that the inhibition rule option sounds interesting. Having an info alert that tells SREs what kubevirt is doing that has the side effect of muting the PDB alert sounds like exactly the behavior we're looking for.

Reply:
I am not sure if console can handle this, so it might show up as not silenced/muted there. Nevertheless even if they do not go with this option, it would be beneficial if they supplied alerts for their workloads/PDBs.